File: sdk/python/aiinfra/_http.py
from typing import Any

import httpx


class AIInfraError(Exception):
    def __init__(self, status_code: int, message: str):
        self.status_code = status_code
        super().__init__(f"[{status_code}] {message}")


class HTTPClient:
    def __init__(
        self,
        base_url: str,
        *,
        admin_token: str | None = None,
        api_key: str | None = None,
        timeout: float = 120.0,
    ):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        headers: dict[str, str] = {}
        if admin_token:
            headers["Authorization"] = f"Bearer {admin_token}"
        elif api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        self._client = httpx.Client(base_url=self.base_url, headers=headers, timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def request(self, method: str, path: str, **kwargs: Any) -> Any:
        resp = self._client.request(method, path, **kwargs)
        if resp.is_error:
            raise AIInfraError(resp.status_code, resp.text)
        if resp.content:
            return resp.json()
        return None

    def get(self, path: str, **kwargs: Any) -> Any:
        return self.request("GET", path, **kwargs)

    def post(self, path: str, **kwargs: Any) -> Any:
        return self.request("POST", path, **kwargs)

    def delete(self, path: str, **kwargs: Any) -> Any:
        return self.request("DELETE", path, **kwargs)

File: sdk/typescript/src/http.ts
export class AIInfraError extends Error {
  constructor(public statusCode: number, message: string) {
    super(`[${statusCode}] ${message}`);
  }
}

export type ClientOptions = {
  adminToken?: string;
  apiKey?: string;
  timeoutMs?: number;
};

export class HttpClient {
  constructor(
    private baseUrl: string,
    private opts: ClientOptions = {},
  ) {}

  private headers(extra?: Record<string, string>): Record<string, string> {
    const h: Record<string, string> = { ...(extra || {}) };
    if (this.opts.adminToken) h.Authorization = `Bearer ${this.opts.adminToken}`;
    else if (this.opts.apiKey) h.Authorization = `Bearer ${this.opts.apiKey}`;
    return h;
  }

  async request<T>(method: string, path: string, init?: RequestInit): Promise<T> {
    const resp = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: { ...this.headers(init?.headers as Record<string, string>), ...init?.headers },
      signal: AbortSignal.timeout(this.opts.timeoutMs ?? 120000),
    });
    const text = await resp.text();
    if (!resp.ok) throw new AIInfraError(resp.status, text);
    return text ? JSON.parse(text) : (null as T);
  }

  get<T>(path: string) {
    return this.request<T>('GET', path);
  }

  post<T>(path: string, body?: unknown) {
    return this.request<T>('POST', path, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: body ? JSON.stringify(body) : undefined,
    });
  }

  delete<T>(path: string) {
    return this.request<T>('DELETE', path, { method: 'DELETE' });
  }

  put<T>(path: string, body?: unknown) {
    return this.request<T>('PUT', path, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: body ? JSON.stringify(body) : undefined,
    });
  }
}
